docs/user/performance.rst

```diff
@@ -0,0 +1,122 @@
+===========
+Performance
+===========
+
+Since skorch provides a few extras on top of a pure PyTorch training code, it is
+expected that it will add an overhead to the total runtime. For typical
+workloads, this overhead should be unnoticeable.
+
+In a few situations, the skorch extras may add a significant overhead, though.
+This is especially the case when the amount of data and the neural net are
+relatively small. The reason is that typically, most time is spent on the
+``forward``, ``backward``, and parameter update calls. When those are really
+fast, the skorch overhead will get noticed.
+
+There are, however, a few things that can be done to reduce the skorch overhead.
+We will focus on accelerating the training process, where the overhead should be
+largest. Below, some mitigations are described, including the potential
+downsides.
+
+First make sure that there is any significant slowdown
+------------------------------------------------------
+
+Neural nets are notoriously slow to train. Therefore, if your training takes a
+lot of time, that doesn't automatically mean that the skorch overhead is at
+fault. Maybe the training would take the same time without skorch. If you have
+some measurements about training the same model without skorch, first make sure
+that this points to skorch being the culprit before trying to optimize using the
+mitigations described below. If it turns out skorch is not the culprit, look
+into optimizing the performance of PyTorch code in general.
+
+Many people use skorch for hyper-parameter search. Remember that this implies
+fitting the model repeatedly, thus a long run time is expected. E.g. if you run
+a grid search on two hyper-parameters, each with 10 variants, and 5 splits,
+there will actually be 10 x 10 x 5 fit calls, so expect the process to take
+approximately 500 times as long as a single model fit. Increase the verbosity on
+the grid search to get a better idea on the progress (e.g. ``GridSerachCV(...,
+verbose=3)``).
+
+Turning off verbosity
+---------------------
+
+By default, skorch produces a print log of the training progress. This is useful
+for checking the training progress, monitor overfitting, etc. If you don't need
+these diagnostics, you can turn them off via the ``verbose`` parameter. This
+way, printing is deactivated, saving time on i/o. You can still access the
+diagnostics through the ``history`` attribute after training has finished.
```

Comment:
From https://github.com/skorch-dev/skorch/pull/730#issuecomment-748932720, I am surprised that setting `verbose=0` makes the time go from 3.0 to 2.1. We do not really use `self.verbose` in many places.

Reply:
> We do not really use self.verbose in many places.

One idea with `verbose` is that users can easily write their own callbacks with some helpful debugging print calls that they may turn off with a single parameter. Whether many users do, I can't say.

> I am surprised that setting verbose=0 makes the time go from 3.0 to 2.1

My guess with this specific test is that the overhead depends on the speed at which the environment flushes the print. Not sure if it's blocking or not. I.e. if you have a faster terminal, the overhead could be less. But it's just a guess.